src/connectionbenchmark/connectionbenchmark.cpp

```diff
@@ -0,0 +1,142 @@
+
+  emit speedChanged();
+  setState(StateReady);
+}
+
+void ConnectionBenchmark::setState(State state) {
+  logger.debug() << "Set state" << state;
+  m_state = state;
+
+  emit stateChanged();
+}
+
+void ConnectionBenchmark::start() {
+  logger.debug() << "Start connection benchmarking";
+
+  Q_ASSERT(m_state != StateRunning);
+
+  MozillaVPN* vpn = MozillaVPN::instance();
+  Q_ASSERT(vpn);
+
+  Controller* controller = vpn->controller();
+  Controller::State controllerState = controller->state();
+  Q_ASSERT(controllerState == Controller::StateOn);
+
+  m_connectionHealth = vpn->connectionHealth();
+  if (m_connectionHealth->stability() == ConnectionHealth::NoSignal) {
+    handleStabilityChange();
+  }
+
+  connect(controller, &Controller::stateChanged, this,
+          &ConnectionBenchmark::stop);
+  connect(m_connectionHealth, &ConnectionHealth::stabilityChanged, this,
+          &ConnectionBenchmark::handleStabilityChange);
+
+  // Create ping benchmark
+  BenchmarkTaskPing* pingTask = new BenchmarkTaskPing();
+  connect(pingTask, &BenchmarkTaskPing::finished, this,
+          &ConnectionBenchmark::pingBenchmarked);
+  connect(pingTask, &Task::completed, this,
+          [this, pingTask]() { m_benchmarkTasks.removeOne(pingTask); });
+  m_benchmarkTasks.append(pingTask);
+  TaskScheduler::scheduleTask(pingTask);
+
+  // Create download benchmark
+  BenchmarkTaskDownload* downloadTask =
+      new BenchmarkTaskDownload(Constants::BENCHMARK_DOWNLOAD_URL);
+  connect(downloadTask, &BenchmarkTaskDownload::finished, this,
+          &ConnectionBenchmark::downloadBenchmarked);
+  connect(downloadTask, &Task::completed, this,
+          [this, downloadTask]() { m_benchmarkTasks.removeOne(downloadTask); });
+  m_benchmarkTasks.append(downloadTask);
+  TaskScheduler::scheduleTask(downloadTask);
+
+  setState(StateRunning);
+}
+
+void ConnectionBenchmark::stop() {
+  logger.debug() << "Stop benchmark";
+
+  if (m_state == StateRunning || m_state == StateError) {
+    Q_ASSERT(!m_benchmarkTasks.isEmpty());
+
+    for (BenchmarkTask* benchmark : m_benchmarkTasks) {
+      benchmark->stop();
+    }
+
+    m_benchmarkTasks.clear();
+  };
+
+  if (m_state != StateError) {
+    setState(StateInitial);
+  }
+}
+
+void ConnectionBenchmark::downloadBenchmarked(quint64 bytesPerSecond,
+                                              bool hasUnexpectedError) {
+  logger.debug() << "Benchmarked download" << bytesPerSecond;
+
+  if (hasUnexpectedError) {
+    setState(StateError);
+    return;
+  }
+
+  m_download = bytesPerSecond;
+  emit downloadChanged();
+
+  setConnectionSpeed();
+}
+
+void ConnectionBenchmark::pingBenchmarked(quint64 pingLatency) {
+  logger.debug() << "Benchmarked ping" << pingLatency;
+
+  m_ping = pingLatency;
+  emit pingChanged();
+}
+
+void ConnectionBenchmark::handleStabilityChange() {
+  logger.debug() << "Handle stability change";
+
+  if (m_connectionHealth->stability() == ConnectionHealth::NoSignal) {
+    setState(StateError);
```

Comment:
ah! ok! so... in case there is "no-signal". how do we recover from it?

Reply:
If there is an error we’ll show an error screen ([Figma spec](https://www.figma.com/file/3V7aahpQ096Qsp2L3SCT6h/Connection-Screen%2FGraph-Redesign?node-id=504%3A4005)) that I’ll still need to be added.

Reply:
Right, there is no way to recover currently. I’ll add it!

Reply:
I added a `reset` method that can be used to abort the benchmark or recover from an error from the frontend.